rmw_zenoh_cpp/src/detail/graph_cache.cpp

```diff
@@ -187,29 +189,55 @@ void GraphCache::parse_put(const std::string & keyexpr)
   }
 
   // Add the node to the namespace if it did not exist and return.
-  NodeMap::iterator node_it = ns_it->second.find(entity.node_name());
-  if (node_it == ns_it->second.end()) {
-    ns_it->second.insert(std::make_pair(entity.node_name(), make_graph_node(entity)));
-    RCUTILS_LOG_WARN_NAMED(
-      "rmw_zenoh_cpp", "Added node /%s to an existing namespace %s in the graph.",
-      entity.node_name().c_str(),
-      entity.node_namespace().c_str());
-    return;
+  // Case 1: First time a node with this name is added to the namespace.
+  // Case 2: There are one or more nodes with the same name but the entity could
+  // represent a node with the same name but a unique id which would make it a
+  // new addition to the graph.
+  std::pair<NodeMap::iterator, NodeMap::iterator> range = ns_it->second.equal_range(
+    entity.node_name());
+  NodeMap::iterator node_it = std::find_if(
+    range.first, range.second,
+    [&entity](const std::pair<std::string, GraphNodePtr> & node_it)
+    {
+      return entity.id() == node_it.second->id_;
+    });
+  if (node_it == range.second) {
+    // Either the first time a node with this name is added or with an existing
+    // name but unique id.
+    NodeMap::iterator insertion_it =
+      ns_it->second.insert(std::make_pair(entity.node_name(), make_graph_node(entity)));
+    if (insertion_it != ns_it->second.end()) {
+      RCUTILS_LOG_WARN_NAMED(
+        "rmw_zenoh_cpp",
+        "Added a new node /%s with id %s to an existing namespace %s in the graph.",
+        entity.node_name().c_str(),
+        entity.id().c_str(),
+        entity.node_namespace().c_str());
+      return;
+    } else {
+      RCUTILS_LOG_ERROR_NAMED(
+        "rmw_zenoh_cpp",
+        "Unable to add a new node /%s with id %s an existing namespace %s in the graph. Report this bug.",
+        entity.node_name().c_str(),
+        entity.id().c_str(),
+        entity.node_namespace().c_str());
+      return;
+    }
+  } else {
+    // The entity represents a node that already exists in the graph.
+    // Update topic info if required below.
   }
```

Comment:
I think we can simplify this whole block a bit:

```suggestion
    if (insertion_it != ns_it->second.end()) {
      RCUTILS_LOG_INFO_NAMED(
        "rmw_zenoh_cpp",
        "Added a new node /%s with id %s to an existing namespace %s in the graph.",
        entity.node_name().c_str(),
        entity.id().c_str(),
        entity.node_namespace().c_str());
    } else {
      RCUTILS_LOG_ERROR_NAMED(
        "rmw_zenoh_cpp",
        "Unable to add a new node /%s with id %s an existing namespace %s in the graph. Report this bug.",
        entity.node_name().c_str(),
        entity.id().c_str(),
        entity.node_namespace().c_str());
    }
    return;
  }
  // Otherwise, the entity represents a node that already exists in the graph.
  // Update topic info if required below.
```